lib/time_schedule_parser.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require 'active_support/core_ext/numeric/time'
+require 'tzinfo'
+
+# Parse time definition strings for validity
+class TimeScheduleParser
+  attr_reader :weekdays
+
+  TimeSchedule = Struct.new(:day_from, :day_to, :hour_from, :minute_from, :hour_to, :minute_to, :timezone)
+
+  def initialize
+    @weekdays = %w[MON TUE WED THU FRI SAT SUN]
+  end
+
+  def time_schedule(opts)
+    TimeSchedule.new(opts[:day_from], opts[:day_to], opts[:hour_from], opts[:minute_from], opts[:hour_to], opts[:minute_to], opts[:timezone])
+  end
+
+  def parse_schedule(schedule)
+    # Example: "Mon-Fri 09:00-17:30 Europe/London"
+    regexp = %r{^([a-zA-Z]{3})-([a-zA-Z]{3}) (\d\d):(\d\d)-(\d\d):(\d\d) ([a-zA-Z/_]+)$}
+    matches = schedule.match(regexp)
+
+    raise 'Schedule does not match regex' if matches.nil?
+    raise "Timezone is invalid: '#{matches.captures[6]}'" unless TZInfo::Timezone.all_identifiers.include?(matches.captures[6])
+
+    day_from = matches.captures[0].upcase
+    day_to = matches.captures[1].upcase
+    raise "Day Range is invalid: '#{day_from}-#{day_to}'" unless ([day_from, day_to] - weekdays).empty?
+
+    time_schedule(
+      day_from: matches.captures[0].upcase,
+      day_to: matches.captures[1].upcase,
+      hour_from: matches.captures[2].to_i,
+      minute_from: matches.captures[3].to_i,
+      hour_to: matches.captures[4].to_i,
+      minute_to: matches.captures[5].to_i,
+      timezone: matches.captures[6]
+    )
+  end
+
+  def schedule_active?(schedule)
+    begin
+      current_time = Time.now.in_time_zone(schedule[:timezone])
+    rescue StandardError
+      return false
```

Comment:
Probably a good idea to log the error here, otherwise it will just be swallowed / hidden.